eng/CodeChecks.ps1

```diff
@@ -0,0 +1,85 @@
+#requires -version 5
+
+$ErrorActionPreference = 'Stop'
+Set-StrictMode -Version 1
+
+$repoRoot = Resolve-Path "$PSScriptRoot/../sdk/"
+
+[string[]] $errors = @()
+
+function LogError([string]$message) {
+    if ($env:TF_BUILD) {
+        Write-Host "##vso[task.logissue type=error]$message"
+    }
+    Write-Host -f Red "error: $message"
+    $script:errors += $message
+}
+
+function Invoke-Block([scriptblock]$cmd) {
+    $cmd | Out-String | Write-Verbose
+    & $cmd
+
+    # Need to check both of these cases for errors as they represent different items
+    # - $?: did the powershell script block throw an error
+    # - $lastexitcode: did a windows command executed by the script block end in error
+    if ((-not $?) -or ($lastexitcode -ne 0)) {
+        if ($error -ne $null)
+        {
+            Write-Warning $error[0]
+        }
+        throw "Command failed to execute: $cmd"
+    }
+}
+
+try {
+
+    Write-Host "Checking that solutions are up to date"
+
+    Get-ChildItem "$repoRoot/Azure.*.sln" -Recurse `
+        | % {
+            Write-Host "  Checking $(Split-Path -Leaf $_)"
+            $slnDir = Split-Path -Parent $_
+            $sln = $_
+            & dotnet sln $_ list `
+                | ? { $_ -ne 'Project(s)' -and $_ -ne '----------' } `
+                | % {
+                        $proj = Join-Path $slnDir $_
+                        if (-not (Test-Path $proj)) {
+                            LogError "Missing project. Solution references a project which does not exist: $proj. [$sln] "
+                        }
+                    }
+        }
+
+    Write-Host "Re-generating readmes"
+    Invoke-Block {
+        & $PSScriptRoot\Update-Snippets.ps1
+    }
+
+    Write-Host "Re-generating clients"
+    Invoke-Block {
+        & $repoRoot\storage\generate.ps1
+    }
+
+    Write-Host "git diff"
+    & git diff --ignore-space-at-eol --exit-code
+    if ($LastExitCode -ne 0) {
+        $status = git status -s | Out-String
+        $status = $status -replace "`n","`n    "
+        LogError "Generated code is not up to date."
```

Comment:
Nit - can you do this check after each section and then add more detailed errors like can you make this `"Generated Storage code is not up to date (please re-run sdk\storage\generate.ps1)."` to make it easy for folks to solve any issues themselves?

Reply:
I've added more info to the message itself, let's see if people would get confused by it and add more granular failures if needed.